Add updating chart with upstream workflow

nickgerace opened the pull request:
Add updating chart with upstream workflow. This includes a
troubleshooting section.

EDIT: you can see what the README looks like rendered [here](https://github.com/nickgerace/rancher-charts/blob/dev-v2.5-source-docs/README.md).

aiyengar2 commented on README.md:
```diff
@@ -98,6 +98,50 @@ The step includes:
+# Now, run prepare. You'll encounter conflicts, but that is okay for now.
+make CHART=<insert-chart> prepare
+
+# Fix the conflict files files manually and store them in a local directory.
+# Store the package.yaml file in the local directory as well/
+mkdir tmp/
+cp <modified-files> tmp/
+cp packages/<insert-chart>/package.yaml tmp/
+
+# Make the repository clean again. You may have to remove artifacts.
+make clean
+git reset --hard
```
I'm not sure if the recommended Bash commands you added to the README to resolve patch conflicts would actually work for all (or most) charts at the moment due to special logic in our scripts to support things like CRD charts or patching subcharts; it would probably be something that works perfectly fine for charts that have simple patches (e.g. longhorn), so maybe we should leave a note about that.

For full coverage, we'd probably need something more heavyweight like https://github.com/aiyengar2/charts/blob/d3e90909c7f28d5c66eb1f287e877c61a58310e8/scripts/generate-rebased-patch, which is a script I used earlier to automate the process of rebasing Rancher Monitoring to a new upstream.

aiyengar2 replied:
For reference, the stuff in https://github.com/aiyengar2/charts/blob/d3e90909c7f28d5c66eb1f287e877c61a58310e8/scripts/generate-rebased-patch#L77-L133 is what actually handled the special cases I mentioned above; this was a script I made for an earlier version of the charts scripts tho, so it probably needs to be rewritten now.

nickgerace replied:
What do you suggest to do with this troubleshooting guide? I needed it for my workflow _because_ the `charts-crd` chart was not created. I couldn't patch my changes because the conflicts caused `charts-crd` to never be generated. The only way to avoid that was to remove the affected files from the patch file.

aiyengar2 replied:
Yeah the charts-crd fails to be created because we patch before we attempt to split out the CRD chart.

I think what you have now is okay since you're just recommending copying the specific files that do have patch conflicts to a temporary directory and removing the conflicting patches from the patch file (as opposed to copying the entire directory, which would probably fail); this should work in general, but probably (not 100% sure) will not work for charts that modify the `requirements.yaml` or try to patch subcharts like monitoring / Istio. It also won't be feasible probably for charts that have enormous patches (again like monitoring / istio).

But neither of those issues impact the majority of chart owners and this README will anyways be rewritten post https://github.com/rancher/rancher/issues/28730, I think this should be good for now 🙂

nickgerace replied:
Fixed by removing the bash commands themselves. Replaced with generalized language that gets the same idea across.

aiyengar2 reviewed (changes requested):
@nickgerace thanks for taking the initiative with this README.md update! Left some comments

nickgerace commented on README.md:
```diff
@@ -98,6 +98,50 @@ The step includes:
+
+To update an existing chart with upstream, run through the following steps.
+
+1. Change the URL in `package.yaml` to the upstream version.
+2. Run through the **Workflow** section above until the `patch` step.
+3. Check the new image versions (and entirely new images) in the upstream chart, and update them in the local chart. There are multiple ways to find them, but the easiest may be to install the upstream chart on a sandboxed cluster and track all images used.
+4. Add or update images in [image-mirror](https://github.com/rancher/image-mirror).
+5. Continue the **Workflow** section at the `patch` step, and complete it.
+6. Once the images have been added/updated, test the newly updated chart.
+
+#### Troubleshoting
```
Sure thing

nickgerace commented on README.md:
```diff
@@ -98,6 +98,50 @@ The step includes:
+# Now, run prepare. You'll encounter conflicts, but that is okay for now.
+make CHART=<insert-chart> prepare
+
+# Fix the conflict files files manually and store them in a local directory.
```
Fixed

nickgerace commented on README.md:
```diff
@@ -98,6 +98,50 @@ The step includes:
+#### Troubleshoting
+
+If upgrading the chart with the upstream version, you may encounter conflicts when running `prepare`.
+Here the recommended workflow in this scenario…
```
I'll do this

nickgerace commented on README.md:
```diff
@@ -98,6 +98,50 @@ The step includes:
+To update an existing chart with upstream, run through the following steps.
+
+1. Change the URL in `package.yaml` to the upstream version.
+2. Run through the **Workflow** section above until the `patch` step.
+3. Check the new image versions (and entirely new images) in the upstream chart, and update them in the local chart. There are multiple ways to find them, but the easiest may be to install the upstream chart on a sandboxed cluster and track all images used.
+4. Add or update images in [image-mirror](https://github.com/rancher/image-mirror).
+5. Continue the **Workflow** section at the `patch` step, and complete it.
```
Okay that sounds good

nickgerace commented:
Addressed all your comments but one @aiyengar2. Thank you for taking a look!

Even if it's not the `bash` commands, I think that having *some* form of documentation on what to do in a conflict scenario is important though. I only knew to cache my changes in a `tmp/` directory and edit the patch file manually through word of mouth. However, that could also be because I am new to the patch workflow.

nickgerace commented:
Addressed @aiyengar2's last review by removing the bash commands themselves. Replaced with generalized language that gets the same idea across.

aiyengar2 commented on README.md:
```diff
@@ -98,6 +98,26 @@ The step includes:
 
+### Troubleshooting
+
+When updating your chart with the upstream version, you may encounter conflicts, or have general questions about the process.
+Below are some troubleshooting tips.
```
nit:
```suggestion
When updating your chart with the upstream version, you may encounter conflicts or have general questions about the process. Below are some troubleshooting tips:
```

aiyengar2 reviewed (approved):
nit comment, LGTM otherwise

rmweir commented on README.md:
```diff
@@ -98,6 +98,26 @@ The step includes:
 
+### Troubleshooting
+
+When updating your chart with the upstream version, you may encounter conflicts, or have general questions about the process.
```
nit: I'm not sure if the first comma is necessary

rmweir commented on README.md:
```diff
@@ -98,6 +98,26 @@ The step includes:
+When updating your chart with the upstream version, you may encounter conflicts, or have general questions about the process.
+Below are some troubleshooting tips.
+
+#### How do I find the all image tags from the upstream chart?
```
Should this be
```suggestion
#### How do I find all the image tags from the upstream chart?
```

aiyengar2 reviewed (approved):
LGTM

rmweir reviewed (approved):
LGTM
